sympy/core/equation.py

```diff
@@ -0,0 +1,508 @@
+    >>> Eqn(Integral(q.lhs,b),integrate(q,b,side='rhs'))
+    Integral(a*c, b) = b**2/(2*c)
+
+    SymPy's solvers do not understand these equations. They expect an
+    expression that the solver assumes = 0. Thus to use the solver the
+    equation must be rearranged so that all non-zero symbols are on one side.
+    Then just the non-zero symbolic side is passed to ``solve()``.
+    >>> t2 = t-t.rhs
+    >>> t2
+    a - b/c = 0
+    >>> solve(t2.lhs,c)
+    [b/a]
+    """
+
+    def __new__(cls, lhs, rhs, **kwargs):
+        check = kwargs.pop('check', True)
+        lhs = _sympify(lhs)
+        rhs = _sympify(rhs)
+        if not isinstance(lhs,Expr) or not isinstance(rhs,Expr):
+            raise TypeError('lhs and rhs must be valid sympy expressions.')
+        if check and lhs.is_number and rhs.is_number:
+            tst=(lhs-rhs)
+            if tst.is_number:
+                print('Lhs & rhs are numbers. Checking statement validity.'
+                      'Please be patient. This can be slow for complex '
+                      'expressions...')
+                tst=tst.simplify().evalf()
+            if tst != 0 and tst.is_number:
+                from warnings import warn
+                warnstr = '\nDid you really mean to define unequal '
+                warnstr += 'numbers '
+                warnstr += str(lhs) + ' and ' + str(rhs)
+                warnstr += ' as equal?\n'
+                warnstr += 'To suppress this warning include `check=False`'
+                warnstr += ' in the equation definition: '
+                warnstr += '`Eqn(lhs,rhs, check=False)`.'
+                warn(warnstr)
+        return super().__new__(cls, lhs, rhs)
+
+    @property
+    def lhs(self):
+        """
+        Returns the lhs of the equation.
+        """
+        return self.args[0]
+
+    @property
+    def rhs(self):
+        """
+        Returns the rhs of the equation.
+        """
+        return self.args[1]
+
+    def as_Boolean(self):
+        """
+        Converts the equation to an Equality.
+        """
+        from .relational import Equality
+        return Equality(self.lhs, self.rhs)
+
+    @property
+    def reversed(self):
+        """
+        Swaps the lhs and the rhs.
+        """
+        return Equation(self.rhs, self.lhs, check=False)
+
+    @property
+    def swap(self):
+        """
+        Synonym for `.reversed`
+        """
+        return self.reversed
+
+    def _applytoexpr(self, expr, func, *args, **kwargs):
+        # Applies a function to an expression checking whether there
+        # is a specialized version associated with the particular type of
+        # expression. Errors will be raised if the function cannot be
+        # applied to an expression.
+        funcname = getattr(func, '__name__')
+        if hasattr(expr, funcname):
+            return getattr(expr, funcname)(*args, **kwargs)
+        else:
+            return func(expr, *args, **kwargs)
```

Comment:
Rather than checking `hasattr` and then using `getattr` you can use the 3-arg form of `getattr`:
```
func = getattr(expr, funcname, None)
if func is not None:
    func(...)
```
This avoids a double lookup (`hasattr` internally calls `getattr`).

Reply:
Was unaware of that option. Good catch!

Reply:
Changed to:
```
        funcname = getattr(func, '__name__')
        localfunc = getattr(expr, funcname, None)
        if localfunc is not None:
            return localfunc(*args, **kwargs)
        return func(expr, *args, **kwargs)
```
commit d60735a9